tests/Integration/Behaviour/Features/Context/Domain/CategoryFeatureContext.php

```diff
@@ -285,8 +270,190 @@ public function editRootCategoryWithFollowingDetails(string $rootCategory, Table
         $command->setAssociatedGroupIds($editableRootCategoryTestCaseData->getGroupAssociationIds());
 
         $this->getCommandBus()->handle($command);
+    }
+
+    /**
+     * @When I add new root category :categoryReference with following details:
+     *
+     * @param string $categoryReference
+     * @param TableNode $table
+     */
+    public function addNewRootCategoryWithFollowingDetails(string $categoryReference, TableNode $table)
+    {
+        $testCaseData = $table->getRowsHash();
+        $editableRootCategoryTestCaseData = $this->mapDataToEditableCategory($testCaseData);
+
+        /** @var EditCategoryCommand $command */
+        $command = new AddRootCategoryCommand(
```

Comment:
@matks - general question related to command. Would be interesting to know why we are passing primitive types to commands constructors instead of objects and entities? Looking far from now - if one day we will change object model to doctrine - then we'll have lots of refactoring once again. Also Symfony forms are capable of handling into objects/entities (possibility of less code).

What part of the PrestaShop refactoring puzzle am I missing?

Reply:
It's intended 😊this is one of CQRS rules.
Commands and Queries must contain only primitive types, although a few exceptions are possible but only bu providing custom immutable types (ex: `Money`).

The idea is that Commands and Queries represent an intention of the user: `ViewUser`, `AddOrderFromBO`, `DisableWarehouse`. These intentions are constructed by a user using the types he has: strings, integers, booleans, usually passed by HTTP requests.

As https://stackoverflow.com/questions/15891143/cqrs-command-constructs states:
"A command is data-oriented, not object-oriented. Its properties should be mostly primitives. You're not modeling anything here, just sending flat data around."

And all these types are immutable which helps making sure a Command or Query is immutable too.

Moreover we want to make dispatching Commands and Queries as easy as possible for developers so using primitives only make it very easy to build one.

Reply:
Thank you for explaining to me that. :)